fplll/pruner/pruner.h

```diff
@@ -290,37 +382,186 @@ template <class FT> class Pruner
       symmetry_factor = 1;
     }
     min_pruning_coefficients.resize(d);
+    btmp.resize(d);
+    bftmp.resize(n);
     fill(min_pruning_coefficients.begin(), min_pruning_coefficients.end(), 0.);
-
     set_tabulated_consts();
+    // need to fix target if possible
+    if (metric == PRUNER_METRIC_PROBABILITY_OF_SHORTEST)
+    {
+      // if target > 1 or < 0, optimize overall cost, use
+      //   0.99 as target for the computation in repeated_enum_cost().
+      if (this->target > 1.0 || this->target < 0.0)
+      {
+        this->target = 0.99;
+        opt_overall  = true;
+      }
+    }
+    else if (metric == PRUNER_METRIC_EXPECTED_SOLUTIONS)
+    {
+      // if target < 0, optimize overall cost, use
+      //   0.99 as target for the computation in repeated_enum_cost().
+      if (this->target < 0.0)
+      {
+        this->target = 0.99;
+        opt_overall  = true;
+      }
+    }
+    else
+    {
+      throw std::invalid_argument("Pruner was set to an unknown metric");
+    }
     load_basis_shapes(gso_rs);
   }
 
   /**
+     @brief run the optimization process using 'even' coefficients. Note
+     the optimization only applies to the pruning coefficients indexed
+     by (0, 2, 4, ... n). It thus uses half of the coefficients.
+
      Run the optimization process, successively using the algorithm activated
+     using using half coefficients.
 
-     @brief run the optimization process
+     Note the input pr has length n; but only the even indices in the vector
+     will be used in the optimization. In the end, we have pr_i = pr_{i+1}.
   */
-  void optimize_coefficients(/*io*/ vector<double> &pr);
+  void optimize_coefficients_evec(/*io*/ vector<double> &pr);
+
+  /**
+     @brief run the optimization process using all the coefficients.
+
+     Run the optimization process, successively using the algorithm activated
+     using using full coefficients.
+
+     Note the input pr has length n.
+  */
+  void optimize_coefficients_full(/*io*/ vector<double> &pr);
+
+  /**
+     @brief tune the pruning parameter to reduce single enumeration time
+
+     Optimization process to the pruning parameters to reduce single enumeration
+     time with the hope that it also reduces the overall enumeration time.
+  */
+  void optimize_coefficients_tune_single(/*io*/ vector<double> &pr);
+
+  /**
+     @brief tune the pruning parameter to increase succ. probability
+
+     Optimization process to the pruning parameters to increase succ.
+     probability with the restriction that the single enumeration time does
+     not increase significantly.
+  */
+  void optimize_coefficients_tune_prob(/*io*/ vector<double> &pr);
```

Comment:
It is not clear to me why these got a `_tune_` in their name, don't they all tune? Also, is this a helper function used internally or is it at the same level as `optimize_coefficients`?

Reply:
The current optimize_coefficients_cost() and optimize_coefficients_prob() are two main functions. optimize_coefficients_tune_prob() is a helper function used internally.  Do you mean to put all such functions private?

Reply:
In general, I'm in favour of public functions for a library like this as it makes it more hackable. In this case, though, I find these functions confusing, so I'd indeed either make them private or give them names that somehow distinguish better. Perhaps this is only my ignorance talking, though.

Reply:
I made those "tune" functions private. I think it's an English question... I use "tune" to mean "small changes". There are other functions which does optimization but changes "bigger" on the pruning coefficients. I changed their name to "local_tune" and made them private --> let me know if this brings any problem to fpylll.